File: common/src/main/kotlin/net/impleri/playerskills/skills/basic/BasicSkillType.kt
package net.impleri.playerskills.skills.basic

import net.impleri.playerskills.api.Skill
import net.impleri.playerskills.api.SkillType
import net.impleri.playerskills.utils.PlayerSkillsLogger
import net.impleri.playerskills.utils.SkillResourceLocation
import net.minecraft.resources.ResourceLocation
import org.apache.commons.lang3.BooleanUtils

class BasicSkillType : SkillType<Boolean>() {
  override val name: ResourceLocation
    get() = NAME

  override fun castToString(value: Boolean?): String {
    return if (value == null) "" else if (value) STRING_VAL_TRUE else STRING_VAL_FALSE
  }

  override fun castFromString(value: String?): Boolean? {
    return if (value.isNullOrBlank()) null else value == STRING_VAL_TRUE
  }

  override fun can(skill: Skill<Boolean>, expectedValue: Boolean?): Boolean {
    val givenValue = BooleanUtils.toBoolean(skill.value)
    val testValue = expectedValue == null || BooleanUtils.toBoolean(expectedValue)
    PlayerSkillsLogger.SKILLS.debug(
      "Checking if player can ${skill.name} (does $expectedValue->$testValue == $givenValue<-${skill.value})",
    )
    return testValue == givenValue
  }

  override fun getPrevValue(skill: Skill<Boolean>, min: Boolean?, max: Boolean?): Boolean {
    return false
  }

  override fun getNextValue(skill: Skill<Boolean>, min: Boolean?, max: Boolean?): Boolean {
    return true
  }

  companion object {
    val NAME = SkillResourceLocation.of("basic")

    private const val STRING_VAL_TRUE = "true"
    private const val STRING_VAL_FALSE = "false"
  }
}

File: common/src/main/scala/net/impleri/playerskills/skills/basic/BasicSkillType.scala
package net.impleri.playerskills.skills.basic

import net.impleri.playerskills.api.skills.Skill
import net.impleri.playerskills.api.skills.SkillOps
import net.impleri.playerskills.api.skills.SkillType
import net.impleri.playerskills.facades.minecraft.core.ResourceLocation
import net.impleri.playerskills.utils.PlayerSkillsLogger

import scala.util.chaining.scalaUtilChainingOps

case class BasicSkillType(
  override val skillOps: SkillOps = Skill(),
  private val logger: PlayerSkillsLogger = PlayerSkillsLogger.SKILLS,
) extends SkillType[Boolean] {
  override val name: ResourceLocation = BasicSkillType.NAME

  override def castToString(value: Boolean): Option[String] = {
    Option(if (value) BasicSkillType.STRING_TRUE else BasicSkillType.STRING_FALSE)
  }

  override def castFromString(value: String): Option[Boolean] = {
    Option(value == BasicSkillType.STRING_TRUE)
  }

  override def can(skill: Skill[Boolean], threshold: Option[Boolean]): Boolean = {
    (threshold.getOrElse(true) == skill.value.getOrElse(false))
      .tap(
        logger.debugP(
          c => s"Checking if player can ${skill.name} (does $threshold == ${skill.value}? $c)",
        ),
      )
  }

  override def getPrevValue(
    skill: Skill[Boolean],
    min: Option[Boolean] = None,
    max: Option[Boolean] = None,
  ): Option[Boolean] = {
    Option(false)
  }

  override def getNextValue(
    skill: Skill[Boolean],
    min: Option[Boolean] = None,
    max: Option[Boolean] = None,
  ): Option[Boolean] = {
    Option(true)
  }
}

object BasicSkillType {
  val NAME: ResourceLocation = ResourceLocation("basic").get

  private[skills] val STRING_TRUE = "true"
  private[skills] val STRING_FALSE = "true"
}
